# AD_scAtlas

Alzheimerâ€™s disease (AD) is a neurodegenerative disease with complex pathophysiology, and AD-dysregulated pathways are inconsistent across different brain regions and patients. Although single-cell RNA sequencing (scRNA-seq) has been performed in different regions of postmortem AD brains, the common and distinct molecular features among different regions remains largely unclear. This hinders the discovery of repurposable and personalized drugs for AD. We combined four scRNA-seq datasets and systematically investigated the common and distinct cellular responses, cell subpopulations, and transcription factors involved in AD. Moreover, we explored the transcriptional heterogeneity of different AD subtypes at the single-cell level. Finally, we conducted individual-based drug repurposing analysis to explore repurposable and personalized drugs. Six major brain cell types were detected after scRNA-seq batch-effect removal and noise cells filtering. Integration with genome-wide association studies (GWAS) summary statistics demonstrated that AD-susceptible genes were mainly enriched with differentially expressed genes (DEGs) in glial cells rather than neuronal cells. While most of DEGs were regulated in opposite directions among different cell types, cell-cell communication analysis revealed several common cellular interaction events involved in neurogenesis, as well as increased cell-cell adhesion. Our comprehensive drug repositioning analysis identified new candidates for AD treatment, including trichostatin, which was predicted to be broadly applicable to different identified AD subtypes, and vorinostat, which was specific for one subtype of AD. In summary, we delineated a cell-specific atlas of the AD transcriptome. Our work illustrated strong cellular heterogeneity in AD for defining AD subtypes. The cell-specific features are important for understanding AD etiology, progression, and drug discovery.

gitub.com/LPH-BIG/scDrug
